### DISCUSSION TITLE
Errors arise when employing boundary conditions

### DISCUSSION THREAD

**Discussion by DeclanDaly** [bfieldtools/bfieldtools#61] [Q&A] [answered]
I am trying to make a set of biplanar coils that have the following set of boundary conditions:
z-coils: at center, $B_z = 0, \frac{dB_z}{dz}= c$
x, y-coils: at center, $B_z = 0, \frac{dB_z}{dx, y}= c$

I have made a program that shows what these field lines look like: 
![image](https://github.com/bfieldtools/bfieldtools/assets/140729826/24e1eafd-9aa6-4d88-a896-a889456af568)

When trying to use bfieldtools to simulate coils for these conditions, I replace the block 
```
target_field = np.zeros(target_points.shape)
target_field[:, 0] += 1
```
with
```
def TargetBFunc(points, c_val, flag):
    target_field=[]
    for elem in points:
        zval = -c_val*elem[1]
        xyval = c_val*elem[2]
        match flag:
            case 'x':
                target_field.append([xyval, 0, 0])
            case 'y':
                target_field.append([0, xyval, 0])
            case 'z':
                target_field.append([0, zval, 0])
    target_field=np.asarray(target_field)
    target_field = target_field / np.max(target_field)
    return np.asarray(target_field)

target_field=TargetBFunc(target_points, 1, 'z')
```
very often encounters the following error:
```
Traceback (most recent call last):

  File ~\anaconda3\Lib\site-packages\spyder_kernels\py3compat.py:356 in compat_exec
    exec(code, globals, locals)

  File c:\users\MYUSER\downloads\biplanar_coil_design.py:153
    coil.s, prob = optimize_streamfunctions(

  File ~\anaconda3\Lib\site-packages\bfieldtools\coil_optimize.py:308 in optimize_streamfunctions
    problem.variables()[0].value / s[0], mesh_conductor=mesh_conductor

TypeError: unsupported operand type(s) for /: 'NoneType' and 'float'
```
I saw a different GitHub discussion had a similar error to me: [https://github.com/bfieldtools/bfieldtools/discussions/51](https://github.com/bfieldtools/bfieldtools/discussions/51)

which motivated the normalization line
`target_field = target_field / np.max(target_field)`

How do I resolve this error for my field specifications?  Or, is there a different method I should be considering?

**Comment by rzetter** [comment, marked as answer]
Hi!  From a quick first glance, I'd say that your target is physically impossible due to not taking concomitant gradients into account.

The Jacobian matrix of the magnetic field is always symmetric (due to Maxwell), and its determinant is always zero. Thus, while the Jacobian of the magnetic field is a 3 x 3 matrix, there is only 5 degrees of freedom for the gradients. As a consequence one cannot create a gradient in one direction without simultaneously creating a so-called concomitant gradient in another direction. 

When creating gradients, it may be convenient to set the target field specification via the use of spherical harmonics coefficients, which capture this behaviour. There's several examples included in the bfieldtools package, e.g. [here](https://bfieldtools.github.io/auto_examples/coil_design/spherical_harmonics_coil_design.html#sphx-glr-auto-examples-coil-design-spherical-harmonics-coil-design-py). You can either use them to simply get the target field values or even optimize the stream functions fully in a spherical-harmonics framework.

PS. The line which throws the error you get `problem.variables()[0].value / s[0]`, is because the solver doesn't converge (probably some error print above the traceback). The vector `s` is the singular value vector of the input data, and before passing the data to the solver the data is scaled to a numerically convenient scale by multiplying by the largest singular value. Afterwards, the data is returned to the original scale by dividing by the largest singular value.
